---
title: "Class15: Pathway Analysis from RNA-Seq Results"
author: "Belinda Xue"
date: "11/14/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## About our Input Data
The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

> Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

## Import count and metadata
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metada and take a peak
colData = read.csv(metaFile, row.name=1)
head(colData)
```

```{r}
# import count and take a peak
countData = read.csv(countFile, row.names=1)
head(countData)
```

Q. Complete the code below to remove the troublesome first column from countData

The **length** colum in countData is going to cause problems with the requried matching to the metadata file so lets remove it here 

```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```

Double check that the colnames in countData match the id values in the colData metadata file
```{r}
colnames(countData)
```

```{r}
rownames(colData)
```

We can use all() function to check if all entries in a vector are TRUE
```{r}
all ( colnames(countData) == rownames(colData) )
```

```{r}
# test how the all() function works; should give you FALSE cuz not all the things in the vector is the same
all( c(T,F,T))
```

## Remove zero count genes

We want to remove genes that have 0 count values in all experiments (i.e. rows that have 0 across all genes)


Q. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns).

Tip: What will rowSums() of countData return and how could you use it in this context?


!= means anything that is not equal to zero, want to keep the one that is not equal to 0 
```{r}
# Filter count data where you have 0 read count across all samples.
countData = countData [ rowSums(countData) != 0,  ]
head(countData)
```

## DESeq2 analysis

```{r}
library(DESeq2)
```



```{r}
# setup the object with our data in the way DEseq want it
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
# Run the analysis
dds = DESeq(dds)
```

Get our results... Give us log2FoldChange, p-value, adjusted p-value... 
```{r}
res = results(dds)
res
```


## Volcono plot

Now we will make a volcano plot, a commonly produced visualization from this type of data that we introduced last day. Basically it's a plot of log2 fold change vs -log adjusted p-value.

```{r}
plot(res$log2FoldChange, -log(res$padj))
```



Lets add some color
```{r}
# Make a color vector for all genes
mycols <- rep("gray", length(res$padj))

# Color blue the genes with absolute fold change above 2
mycols [abs (res$log2FoldChange) > 2] <- "blue"


mycols [res$padj > 0.005] <- "gray"

plot(res$log2FoldChange, -log(res$padj), col=mycols)


```



## Adding gene annotation

Since we mapped and counted against the Ensembl annotation, our results only have information about Ensembl gene IDs. However, our pathway analysis downstream will use KEGG pathways, and genes in KEGG pathways are annotated with Entrez gene IDs. So lets add them as we did the last day.


Add gene symbols and entrez ids / translate between diff lanuages
```{r eval=FALSE}
BiocManager::install("AnnotationDbi")
BiocManager::install("org.Hs.eg.db") # gene annotation
```


```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

# we can traslate between all the following database ID systems
columns(org.Hs.eg.db)
```

Q. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME annotation to our results by completing the code below.

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(countData), # where are my IDS
                    keytype="ENSEMBL", # what format are my IDS
                    column="SYMBOL", # the new format I want
                    multiVals="first")

res
```

```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(countData), # taking number from countData
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res
```


```{r}
res$name =  mapIds(org.Hs.eg.db,
                    keys=row.names(res), # pulling names from the residue name, not from the countdata cuz countdata is all numerical matrix
                    keytype="ENSEMBL", 
                    column="GENENAME",
                    multiVals="first")
res

```



## Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.


```{r}
library(pathview)
library(gage)
library(gageData)

```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs) # singaling metabolic

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3) 
```


The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

Note that we used the mapIDs() function above to obtain Entrez gene IDs (stored in res$entrez) and we have the fold change results from DESeq2 analysis (stored in res$log2FoldChange).

The main gage() function requires a named vector of fold change, where the nmase of the values are the Entrez gene IDs. 
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez 
head(foldchanges)
```


```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
# taking the gene sets, look for the metabolic biochem pathway in the dataset
```


See help on the gage function with ?gage. Specifically, you might want to try changing the value of same.dir. This value determines whether to test for changes in a gene set toward a single direction (all genes up or down regulated) or changes towards both directions simultaneously (i.e. any genes in the pathway dysregulated). Here, we're using the default same.dir=TRUE, which will give us separate lists for pathways that are upregulated versus pathways that are down-regulated.


Now lets look at the object returned from gage().
```{r}
attributes(keggres)
```


```{r}
# Look at the first few down (less) pathways
# these stats show which pathway is down-regulated based on the p-value
# smaller pathway 
head(keggres$less)
```

To begin with lets manually supply a pathway.id (namely the first part of the "hsa04110 Cell cycle") that we could see from the print out above.
```{r}
# pathway of cell cycle, pathway.id is cell cycle, most are down-regulated... 
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

```{r}
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

Now, let's process our results a bit more to automagicaly pull out the top 5 upregulated pathways, then further process that just to get the pathway IDs needed by the pathview() function. We'll use these KEGG pathway IDs for pathview plotting below.

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

Finally, lets pass these IDs in keggresids to the pathview() function to draw plots for all the top 5 pathways.

```{r}
# view the pathway for these 5 up-regulated gene
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```


## Gene Ontology: focus on biological process

We can also do a similar procedure with gene ontology. Similar to above, go.sets.hs has all GO terms. go.subs.hs is a named list containing indexes for the BP, CC, and MF ontologies. Letâ€™s focus on BP (a.k.a Biological Process) here.


KEGG focuses on metabolic pathway/ biochem part, whereas the GO focuses on the biological process
```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)
```














