---
title: "Untitled"
output: github_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```

Now lets check on GDC status: 
# when is the GDC website updated.. 
```{r}
status()
```

```{r}
projects <- getGDCprojects()
head(projects)
```


If you use the View(projects) function call you can see all the project names (such as Neuroblastoma, Pancreatic Adenocarcinoma, etc.) along with their project IDs (such as TARGET-NBL, TCGA-PAAD, etc.) and associated information.

Moving onto cases() we can use an example from the package associated publication to answer our first from question above (i.e. find the number of cases/patients across different projects within the GDC):

```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

Note that the facet() and aggregations() functions here are from the GenomicDataCommons package and act to group all cases by the project id and then count them up.

If you use the View() function on our new cases_by_project object you will find that the data we are after is accessible via cases_by_project$project.project_id.

Q9. Write the R code to make a barplot of the cases per project. Lets plot this data with a log scale for the y axis (log="y"), rotated axis labels (las=2) and color the bar coresponding to the TCGA-PAAD project


```{r}

x <- cases_by_project$project.project_id

# Make a custom color vector for our plot
colvec <- rep("lightblue", nrow(x))
colvec[ (x$key == "TCGA-PAAD") ] <- "red" # x中的key column, 有没有column有个ID是 TCGA-PAAD
colvec

# Plot with 'log' for y axis and rotate labels with 'las'
par (mar=c(10,4,0,0)) 
barplot(x$doc_count, names.arg=x$key, col=colvec, log="y", las=2)
```


Lets explore some other functions from the related TCGAbiolinks package.

We can use the getSampleFilesSummary() function to determine for a given project how many cases and what type of data we have available for each case:

```{r}
samp <- getSampleFilesSummary("TCGA-PAAD")
head(samp)
```

Now we can use GDCquery() function to focus in on a particular data type that we are interested in. For example, to answer our second question from above - namely ‘find all gene expression data files for all pancreatic cancer patients’:

```{r}
query <- GDCquery(project="TCGA-PAAD",
                  data.category="Transcriptome Profiling",
                  data.type="Gene Expression Quantification")

ans <- getResults(query)
```

```{r}
head(ans)
```


## 4. Variant analysis with R

Note we could go to the NCI-GDC web portal and enter the Advanced Search page and then construct a search query to find MAF format somatic mutation files for our ‘TCGA-PAAD’ project.

After some exploration of the website I came up with the following query: “cases.project.project_id in ["TCGA-PAAD"] and files.data_type in ["Masked Somatic Mutation"] and files.data_format in ["MAF"]”


Q9. How many MAF files for the TCGA-PAAD project were found from this advanced web search?

Lets do the same search in R with the help of the TCGAbiolinks package function GDCquery_Maf(). For brevity we will focus on only one of the MAF files for this project, namely the MuTect2 workflow variant calls.
```{r}
maf.file <- GDCquery_Maf(tumor="PAAD", pipelines = "mutect")
```

```{r}
head(maf.file)
```

### MAF analysis
The MAF file contents is now stored as a dataframe and the maftools package workflow, which starts with a MAF file or dataframe, can proceed, starting with reading the pancreatic cancer MAF file.

```{r}
vars = read.maf(maf = maf.file, verbose = FALSE)

```

Plotting MAF summary.
We can use plotmafSummary() function to plot a summary of the maf object, which displays number of variants in each sample as a stacked barplot and variant types as a boxplot summarized by Variant_Classification. We can add either mean or median line to the stacked barplot to display average/median number of variants across the cohort.

```{r}
plotmafSummary(vars)

```


Drawing oncoplots
A very useful summary representation of this data can be obtained via so-called oncoplots, also known as waterfall plots.

```{r}
oncoplot(maf = vars, top = 10)

```

You might need to run the oncoplot() command in the R Console and then zoom the display to see the full plot (as it is rather large and may not appear initially in your Rmarkdown document before Knitting. Another option is to send your plot to a PNG or PDF plot device directly, for example:


```{r}
# Oncoplot for our top 10 most frequently mutated genes
pdf("oncoplot_panc.pdf")
oncoplot(maf = vars, top = 10, fontSize = 12)
dev.off()
```

Oncostrip
We can visualize any set of genes using the oncostrip() function, which draws mutations in each sample similar to the graphic on the NCI-GDC web portal. Note that oncostrip() can be used to draw any number of genes using the input top or genes arguments

```{r}
oncostrip(maf=vars, genes=c("KRAS", "TP53"))

```

```{r}
lollipopPlot(vars, gene='KRAS')

```


## LAB 18 PART II : Personalized Medicine


Can do it manually
```{r eval=FALSE}

cancer specific sub-sequence (8 before the mutation and 8 after the mutation)
# +8 and -8
SPLPSQAM   D   DLMLSPDD
SPLPSQAM   L   DLMLSPDD

YLDDRNTF   R   HSVVVPYE
YLDDRNTF   V   HSVVVPYE
```

Use R to do the alignment and find the 9-mer --> cancer specific subsequence
```{r}
library(bio3d)
seqs <- read.fasta("lecture18_sequences.fa")
seqs
```


```{r}
## Calculate positional identity scores
ide <- conserv(seqs$ali, method="identity")
ide < 1 # see where is not conserved , find the TRUE cuz that is the mutation spot

```
```{r}
# or
pos <- which(ide < 1) [1]  # tell you the 1st position number
pos
pos -8
pos +8

# or 
subseq.pos <- (pos-8):(pos+8)

seqs$ali[2,subseq.pos]
```


```{r}
# Where as sits not scoreing 1 (i.e. non-identical)
mutant.sites <- which(ide < 1) 

## Exclude gap possitions from analysis
gaps <- gap.inspect(seqs)
mutant.sites <- mutant.sites[mutant.sites %in% gaps$f.inds]

mutant.sites
```


```{r}
## Make a "names" label for our output sequences (one per mutant)
mutant.names <- paste0(seqs$ali["P53_wt",mutant.sites],
                       mutant.sites,
                       seqs$ali["P53_mutant",mutant.sites])

mutant.names

```


```{r}
## Sequence positions surounding each mutant site
start.position <- mutant.sites - 8
end.position <-  mutant.sites + 8

# Blank matrix to store sub-sequences
store.seqs <- matrix("-", nrow=length(mutant.sites), ncol=17)
rownames(store.seqs) <- mutant.names

## Extract each sub-sequence
for(i in 1:length(mutant.sites)) {
  store.seqs[i,] <- seqs$ali["P53_mutant",start.position[i]:end.position[i]]
}

store.seqs
```


```{r}
## Output a FASTA file for further analysis
write.fasta(seqs=store.seqs, ids=mutant.names, file="subsequences.fa")

```























